In [1]:
import sys
# Insert at index 0 so Python checks this folder BEFORE site-packages
sys.path.insert(0, "/home/nfm/ViT-Prisma/src")

DEVICE = 'cuda' # change to cpu if cpu only paradigm
BATCH_SIZE = 16

In [2]:
import json
LOCAL_JSON_PATH = "/home/nfm/ViT-Prisma/demos/imagenet_class_index.json"
with open(LOCAL_JSON_PATH, 'r') as f:
    imagenet_class_index = json.load(f)

wnid_to_name = {}
for idx, (wnid, class_name) in imagenet_class_index.items():
    safe_class_name = class_name.replace(" ", "_").replace("/", "_").replace(",", "")
    wnid_to_name[wnid] = safe_class_name

wnid_to_idx = {wnid: int(idx) for idx, (wnid, name) in imagenet_class_index.items()}
idx_to_name = {int(idx): name for idx, (wnid, name) in imagenet_class_index.items()}

idx_to_wnid = {int(idx): wnid for idx, (wnid, name) in imagenet_class_index.items()}
name_to_idx = {name: int(idx) for idx, (wnid, name) in imagenet_class_index.items()}

In [49]:
from vit_prisma.models.model_loader import load_hooked_model
import torch

# The hf_hub: prefix tells TIMM to download from your specific repo
# model_name = "hf_hub:natihash/vit_base_patch16_clip_224.laion2b_linear_probe"
# model_name = "vit_base_patch16_clip_224.laion2b_ft_in1k"


# model = load_hooked_model(
#     model_name,
#     fold_ln=True,
#     device=DEVICE
# )
model_name = "vit_base_patch16_clip_224.laion2b_ft_in1k"
# model_name = "hf_hub:natihash/vit_base_patch16_clip_224.laion2b_linear_probe"
# model_name = "hf_hub:natihash/vit_base_patch16_clip_224.laion2b_linear_probe_real"
# model_name = "vit_base_patch16_224"

# model_name = "open-clip:laion/CLIP-ViT-B-16-laion2B-s34B-b88K"
# model_name = "hf_hub:natihash/vit_base_patch16_clip_224.laion2b_fullft"

from vit_prisma.models.base_vit import HookedViT
import torch
model = HookedViT.from_pretrained(model_name,
                                        center_writing_weights=True,
                                        center_unembed=True,
                                        fold_ln=True,
                                        refactor_factored_attn_matrices=True,
                                    )

# Force the base model to GPU (just in case)
model = model.to("cuda:0")

# FORCE the Prisma/TransformerLens config to match
model.cfg.device = "cuda:0" 

model.eval()
model.to(DEVICE);
print("Successfully loaded custom linear probe!")

2026-04-29 01:06:08 INFO:root: Model 'vit_base_patch16_clip_224.laion2b_ft_in1k' is supported and passes tests.


*****Loading model 'vit_base_patch16_clip_224.laion2b_ft_in1k' of type 'VISION'


/home/nfm/prisma/lib/python3.10/site-packages/huggingface_hub/file_download.py:949: FutureWarning:

`resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.

2026-04-29 01:06:09 DEBUG:urllib3.connectionpool: Resetting dropped connection: huggingface.co
2026-04-29 01:06:10 DEBUG:urllib3.connectionpool: https://huggingface.co:443 "HEAD /timm/vit_base_patch16_clip_224.laion2b_ft_in1k/resolve/main/config.json HTTP/1.1" 307 0
2026-04-29 01:06:10 DEBUG:urllib3.connectionpool: https://huggingface.co:443 "HEAD /api/resolve-cache/models/timm/vit_base_patch16_clip_224.laion2b_ft_in1k/2b86c7046b92e6b9d2bc3f40068b84b750ca3d48/config.json HTTP/1.1" 200 0
2026-04-29 01:06:13 INFO:timm.models._builder: Loading pretrained weights from Hugging Face hub (timm/vit_base_patch16_clip_224.laion2b_ft_in1k)
2026-04-29 01:06:13 DEBUG:urllib3.connectionpool: https://huggingface.co:443 "HEAD /timm/

Converting the weights of a timm model to a Prisma ViT
LayerNorm folded.
Centered weights writing to residual stream


2026-04-29 01:07:43 INFO:root: Loaded pretrained model vit_base_patch16_clip_224.laion2b_ft_in1k into HookedTransformer


Successfully loaded custom linear probe!


In [50]:
model.head.W_H.shape, model.head.b_H.shape

import torch.nn as nn

# W_H is [768, 512] → in=768, out=512
old_head_module = nn.Linear(old_head.shape[0], old_head.shape[1])  # Linear(768, 512)

with torch.no_grad():
    old_head_module.weight.copy_(old_head.T)  # nn.Linear expects [out, in] so transpose
    old_head_module.bias.copy_(old_bias)       # [512] ✓

model.head = old_head_module.to('cuda:0')

In [51]:
# old_head = model.head.W_H.clone().cpu()
# old_bias = model.head.b_H.clone().cpu()

old_head.shape, old_bias.shape, old_head_module
# model.head.W_H.shape, model.head.b_H.shape

(torch.Size([768, 512]),
 torch.Size([512]),
 Linear(in_features=768, out_features=512, bias=True))

In [6]:
import torch
from torch.utils.data import DataLoader
from torchvision.datasets import ImageFolder
from tqdm import tqdm
import timm

IMAGENET_VAL_DIR = "/home/nfm/data_prisma/imagenet_val/kaggle/input/imagenet-object-localization-challenge/ILSVRC/Data/CLS-LOC/val"
BATCH_SIZE = 256
NUM_WORKERS = 8

data_config = timm.data.resolve_model_data_config(model)
transforms = timm.data.create_transform(**data_config, is_training=False)

dataset = ImageFolder(IMAGENET_VAL_DIR, transform=transforms)

# wnid_to_idx already maps synset → correct timm class index
folder_idx_to_timm_idx = {
    folder_idx: wnid_to_idx[wnid]
    for wnid, folder_idx in dataset.class_to_idx.items()
}

loader = DataLoader(dataset, batch_size=BATCH_SIZE, shuffle=False,
                    num_workers=NUM_WORKERS, pin_memory=True)

model.eval()
top1_correct = top5_correct = total = 0

with torch.no_grad():
    for images, folder_labels in tqdm(loader, desc="Evaluating"):
        images = images.to(DEVICE)

        timm_labels = torch.tensor(
            [folder_idx_to_timm_idx[l.item()] for l in folder_labels],
            device=DEVICE,
        )

        logits = model(images)

        top1_correct += (logits.argmax(dim=1) == timm_labels).sum().item()
        top5_correct += (logits.topk(5, dim=1).indices == timm_labels.unsqueeze(1)).any(dim=1).sum().item()
        total += images.size(0)

print(f"Images   : {total:,}")
print(f"Top-1    : {100 * top1_correct / total:.2f}%")
print(f"Top-5    : {100 * top5_correct / total:.2f}%")

Evaluating: 100%|██████████| 196/196 [02:46<00:00,  1.18it/s]

Images   : 50,000
Top-1    : 80.99%
Top-5    : 95.72%


In [30]:
import argparse
import json
import os
from collections import defaultdict
from pathlib import Path

import numpy as np
import pandas as pd
import torch
import open_clip
from PIL import Image
from tqdm import tqdm


# ---------------------------------------------------------------------------
# Data loading helpers
# ---------------------------------------------------------------------------

def load_from_csv(csv_path: str, max_images: int | None = None):
    """
    Load image paths and captions from the CSV format:
        filepath,title
    Returns:
        image_paths : list[str]  – deduplicated, in stable order
        captions    : list[str]
        img2caps    : dict[str, list[int]]  image_path -> [caption indices]
        cap2img     : dict[int, str]        caption index -> image_path
    """
    df = pd.read_csv(csv_path)
    df.columns = df.columns.str.strip()

    # Deduplicate image paths while preserving order
    seen = {}
    for fp in df["filepath"]:
        if fp not in seen:
            seen[fp] = len(seen)

    if max_images:
        keep_paths = set(list(seen.keys())[:max_images])
        df = df[df["filepath"].isin(keep_paths)]

    image_paths = list(dict.fromkeys(df["filepath"].tolist()))
    captions = df["title"].tolist()
    fp_list = df["filepath"].tolist()

    img2caps: dict[str, list[int]] = defaultdict(list)
    cap2img: dict[int, str] = {}
    for idx, (fp, cap) in enumerate(zip(fp_list, captions)):
        img2caps[fp].append(idx)
        cap2img[idx] = fp

    return image_paths, captions, img2caps, cap2img


def load_from_json(json_path: str, img_dir: str, max_images: int | None = None):
    """
    Load from the official COCO captions_val2017.json annotation file.
    """
    with open(json_path) as f:
        data = json.load(f)

    id2file = {img["id"]: os.path.join(img_dir, img["file_name"])
               for img in data["images"]}

    img2caps: dict[str, list[int]] = defaultdict(list)
    cap2img: dict[int, str] = {}
    captions = []

    for ann in data["annotations"]:
        fp = id2file[ann["image_id"]]
        idx = len(captions)
        captions.append(ann["caption"])
        img2caps[fp].append(idx)
        cap2img[idx] = fp

    image_paths = list(img2caps.keys())
    if max_images:
        image_paths = image_paths[:max_images]
        keep = set(image_paths)
        captions_filtered, cap2img_filtered, img2caps_filtered = [], {}, defaultdict(list)
        new_idx = 0
        for old_idx, cap in enumerate(captions):
            fp = cap2img[old_idx]
            if fp in keep:
                cap2img_filtered[new_idx] = fp
                img2caps_filtered[fp].append(new_idx)
                captions_filtered.append(cap)
                new_idx += 1
        captions = captions_filtered
        cap2img = cap2img_filtered
        img2caps = img2caps_filtered

    return image_paths, captions, img2caps, cap2img


# ---------------------------------------------------------------------------
# Feature extraction
# ---------------------------------------------------------------------------

@torch.no_grad()
def encode_images(model, preprocess, image_paths: list[str], device, batch_size=64):
    feats = []
    for i in tqdm(range(0, len(image_paths), batch_size), desc="Encoding images"):
        batch_paths = image_paths[i: i + batch_size]
        imgs = []
        for p in batch_paths:
            try:
                img = preprocess(Image.open(p).convert("RGB"))
            except Exception as e:
                print(f"Warning: could not load {p}: {e}. Using blank image.")
                img = preprocess(Image.new("RGB", (224, 224)))
            imgs.append(img)
        batch = torch.stack(imgs).to(device)
        # f = model.encode_image(batch)
        f = model(batch)
        f = f / f.norm(dim=-1, keepdim=True)
        feats.append(f.cpu())
    return torch.cat(feats, dim=0)   # (N_images, D)


@torch.no_grad()
def encode_texts(model, tokenizer, captions: list[str], device, batch_size=256):
    feats = []
    for i in tqdm(range(0, len(captions), batch_size), desc="Encoding captions"):
        batch = captions[i: i + batch_size]
        tokens = tokenizer(batch).to(device)
        f = model.encode_text(tokens)
        f = f / f.norm(dim=-1, keepdim=True)
        feats.append(f.cpu())
    return torch.cat(feats, dim=0)   # (N_captions, D)


# ---------------------------------------------------------------------------
# Recall@K computation
# ---------------------------------------------------------------------------

def recall_at_k(
    image_feats: torch.Tensor,   # (N_img, D)
    text_feats:  torch.Tensor,   # (N_txt, D)
    img2caps:    dict[str, list[int]],
    cap2img:     dict[int, str],
    image_paths: list[str],
    ks: tuple[int, ...] = (1, 5, 10),
):
    """
    Compute Recall@K for both retrieval directions.

    Image→Text  (i2t): for each image, rank all texts; hit if any GT caption
                        falls within top-K.
    Text→Image  (t2i): for each caption, rank all images; hit if GT image
                        falls within top-K.
    """
    # Cosine similarity matrix  (N_img × N_txt)
    sim = image_feats @ text_feats.T     # already L2-normalised

    img_idx = {fp: i for i, fp in enumerate(image_paths)}

    # ---- Image → Text ----
    i2t_hits = {k: 0 for k in ks}
    for img_i, fp in enumerate(image_paths):
        gt_cap_indices = img2caps[fp]                    # ground-truth cap ids
        scores = sim[img_i]                              # (N_txt,)
        ranked = scores.argsort(descending=True).tolist()
        for k in ks:
            top_k = set(ranked[:k])
            if any(ci in top_k for ci in gt_cap_indices):
                i2t_hits[k] += 1

    # ---- Text → Image ----
    t2i_hits = {k: 0 for k in ks}
    n_caps = text_feats.shape[0]
    for cap_i in range(n_caps):
        gt_fp = cap2img[cap_i]
        gt_img_i = img_idx[gt_fp]
        scores = sim[:, cap_i]                           # (N_img,)
        ranked = scores.argsort(descending=True).tolist()
        for k in ks:
            if gt_img_i in ranked[:k]:
                t2i_hits[k] += 1

    n_img = len(image_paths)
    i2t = {k: 100.0 * i2t_hits[k] / n_img   for k in ks}
    t2i = {k: 100.0 * t2i_hits[k] / n_caps  for k in ks}
    return i2t, t2i


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

In [44]:
# parser = argparse.ArgumentParser(description="Evaluate CLIP on COCO retrieval (Recall@K)")
# src = parser.add_mutually_exclusive_group(required=True)
# src.add_argument("--csv",  help="Path to CSV file (filepath,title)")
# src.add_argument("--json", help="Path to captions_val2017.json")
# parser.add_argument("--img-dir",   default="",   help="Image directory (used with --json)")
# parser.add_argument("--max-images", type=int, default=None,
#                     help="Limit to first N unique images (e.g. 1000 for fast test)")
# parser.add_argument("--batch-size", type=int, default=64)
# parser.add_argument("--device", default=None,
#                     help="cuda / cpu (auto-detected if not set)")
# args = parser.parse_args()

device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Device: {device}")

# ------------------------------------------------------------------
# Load data
# ------------------------------------------------------------------

image_paths, captions, img2caps, cap2img = load_from_csv(
        "/home/nfm/mycoco/coco_val.csv", max_images=None)

# image_paths, captions, img2caps, cap2img = load_from_json(
#         "/home/nfm/mycoco/captions_val2017.json", "/home/nfm/mycoco/val2017", max_images=1000)

print(f"Images : {len(image_paths)}")
print(f"Captions: {len(captions)}")

# ------------------------------------------------------------------
# Load OpenCLIP ViT-B/16, LAION-2B
# ------------------------------------------------------------------
# print("\nLoading OpenCLIP ViT-B-16 / laion2b_s34b_b88k …")
# newmodel, _, preprocess = open_clip.create_model_and_transforms(
#     "ViT-B-16",
#     pretrained="laion2b_s34b_b88k",   # LAION-2B checkpoint
# )

# from torchvision import transforms

# preprocess = transforms.Compose([
#     transforms.Resize(224, interpolation=transforms.InterpolationMode.BICUBIC),
#     transforms.CenterCrop(224),
#     transforms.ToTensor(),
#     transforms.Normalize(
#         mean=(0.48145466, 0.4578275, 0.40821073),
#         std=(0.26862954, 0.26130258, 0.27577711)
#     )
# ])

tokenizer = open_clip.get_tokenizer("ViT-B-16")
newmodel = newmodel.to(device).eval()

# ------------------------------------------------------------------
# Encode
# ------------------------------------------------------------------
batch_size = 256
image_feats = encode_images(model, preprocess, image_paths, device, batch_size)
text_feats  = encode_texts(newmodel, tokenizer, captions, device, batch_size)

# # ------------------------------------------------------------------
# # Evaluate
# # ------------------------------------------------------------------
# print("\nComputing Recall@K …")
# i2t, t2i = recall_at_k(image_feats, text_feats, img2caps, cap2img, image_paths)

# print("\n" + "=" * 45)
# print(f"  Model : OpenCLIP ViT-B/16  (LAION-2B)")
# print(f"  Images: {len(image_paths)}   Captions: {len(captions)}")
# print("=" * 45)
# print(f"{'Metric':<25} {'R@1':>6} {'R@5':>6} {'R@10':>6}")
# print("-" * 45)
# print(f"{'Image → Text':<25} {i2t[1]:>6.1f} {i2t[5]:>6.1f} {i2t[10]:>6.1f}")
# print(f"{'Text  → Image':<25} {t2i[1]:>6.1f} {t2i[5]:>6.1f} {t2i[10]:>6.1f}")
# mean_r1 = (i2t[1] + t2i[1]) / 2
# print("-" * 45)
# print(f"{'Mean R@1':<25} {mean_r1:>6.1f}")
# print("=" * 45)

2026-04-29 00:49:16 INFO:root: Parsing tokenizer identifier. Schema: None, Identifier: ViT-B-16
2026-04-29 00:49:16 INFO:root: Attempting to load config from built-in: ViT-B-16
2026-04-29 00:49:16 INFO:root: Using default SimpleTokenizer.


Device: cuda
Images : 5000
Captions: 25014


Encoding captions: 100%|██████████| 98/98 [00:20<00:00,  4.69it/s]


In [45]:
def recall_at_k_1k_folds(image_feats, text_feats, img2caps, cap2img,
                          image_paths, ks=(1, 5, 10), n_folds=5, fold_size=1000):
    """Average Recall@K over n_folds random 1K-image subsets (standard protocol)."""
    import random
    rng = random.Random(42)
    indices = list(range(len(image_paths)))

    fold_i2t = {k: [] for k in ks}
    fold_t2i = {k: [] for k in ks}

    for _ in range(n_folds):
        fold_img_indices = rng.sample(indices, fold_size)
        fold_img_paths   = [image_paths[i] for i in fold_img_indices]

        # Collect caption indices that belong to this fold's images
        fold_cap_indices = []
        for fp in fold_img_paths:
            fold_cap_indices.extend(img2caps[fp])

        fold_img_feats = image_feats[fold_img_indices]           # (1000, D)
        fold_txt_feats = text_feats[fold_cap_indices]            # (~5000, D)

        # Remap indices to local scope
        local_img2caps = {fp: [] for fp in fold_img_paths}
        local_cap2img  = {}
        for new_ci, old_ci in enumerate(fold_cap_indices):
            fp = cap2img[old_ci]
            local_img2caps[fp].append(new_ci)
            local_cap2img[new_ci] = fp

        i2t, t2i = recall_at_k(fold_img_feats, fold_txt_feats,
                                local_img2caps, local_cap2img,
                                fold_img_paths, ks=ks)
        for k in ks:
            fold_i2t[k].append(i2t[k])
            fold_t2i[k].append(t2i[k])

    avg_i2t = {k: float(np.mean(fold_i2t[k])) for k in ks}
    avg_t2i = {k: float(np.mean(fold_t2i[k])) for k in ks}
    return avg_i2t, avg_t2i

i2t_1k, t2i_1k = recall_at_k_1k_folds(
    image_feats, text_feats, img2caps, cap2img, image_paths
)
print("--- 1K fold (matches published numbers) ---")
print(f"I2T  R@1/5/10: {i2t_1k[1]:.1f} / {i2t_1k[5]:.1f} / {i2t_1k[10]:.1f}")
print(f"T2I  R@1/5/10: {t2i_1k[1]:.1f} / {t2i_1k[5]:.1f} / {t2i_1k[10]:.1f}")

--- 1K fold (matches published numbers) ---
I2T  R@1/5/10: 63.8 / 86.8 / 93.7
T2I  R@1/5/10: 55.3 / 83.0 / 91.3


org clip
I2T  R@1/5/10: 76.4 / 93.4 / 97.2
T2I  R@1/5/10: 60.4 / 85.6 / 93.0

timm finetuned
I2T  R@1/5/10: 28.3 / 55.2 / 68.2
T2I  R@1/5/10: 17.7 / 39.3 / 51.4

myfinetuned
I2T  R@1/5/10: 63.8 / 86.8 / 93.7
T2I  R@1/5/10: 55.3 / 83.0 / 91.3

In [46]:
newmodel.logit_scale.exp()

tensor(100., device='cuda:0', grad_fn=<ExpBackward0>)

In [52]:
import torch
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchvision.datasets import ImageFolder
from tqdm import tqdm
import open_clip

# ── Assumes you already have these loaded ────────────────────────────────────
# model       : open_clip model
# tokenizer   : open_clip tokenizer  
# preprocess  : open_clip image transform
# e.g.:
# model, _, preprocess = open_clip.create_model_and_transforms('ViT-B-16', pretrained='laion2b_s34b_b88k')
# tokenizer = open_clip.get_tokenizer('ViT-B-16')

# model_name = "ViT-B-16"
# pretrained = "laion2b_s34b_b88k"
# newmodel, _, preprocess = open_clip.create_model_and_transforms(model_name, pretrained=pretrained)
# tokenizer = open_clip.get_tokenizer(model_name)

# newmodel.eval()
# newmodel.to(DEVICE)


# ── 1. ImageNet zero-shot prompt templates (from original CLIP paper) ────────
IMAGENET_TEMPLATES = [
    'a photo of a {}.',
    'a blurry photo of a {}.',
    'a photo of many {}.',
    'a photo of the large {}.',
    'a photo of the small {}.',
    'a bad photo of a {}.',
    'a good photo of a {}.',
    'a rendering of a {}.',
    'a cropped photo of a {}.',
    'a jpeg corrupted photo of a {}.',
    'itap of a {}.',
    'itap of my {}.',
    'itap of the {}.',
    'a origami {}.',
    'a toy {}.',
    'a tattoo of a {}.',
    'a embroidered {}.',
    'a plastic {}.',
    'a dark photo of a {}.',
    'a bright photo of a {}.',
    'a photo of a dirty {}.',
    'a photo of a cool {}.',
    'a photo of a weird {}.',
    'art of the {}.',
    'a photo of one {}.',
    'a photo of my {}.',
    'a close-up photo of a {}.',
    'a black and white photo of the {}.',
    'a painting of the {}.',
    'a painting of a {}.',
    'a pixelated photo of the {}.',
    'a sculpture of the {}.',
    'a bright photo of the {}.',
    'a cropped photo of the {}.',
    'a good photo of the {}.',
    'a close-up photo of the {}.',
    'a rendition of the {}.',
    'a rendition of a {}.',
]

# ── 2. Build zero-shot text classifier weights ────────────────────────────────
@torch.no_grad()
def build_zero_shot_classifier(model, tokenizer, class_names, templates, device):
    """
    Returns classifier weights of shape (num_classes, embed_dim),
    one averaged+normalised embedding per class across all templates.
    """
    classifier_weights = []
    for class_name in tqdm(class_names, desc="Building text classifier"):
        # encode every template for this class, then average
        texts = tokenizer([t.format(class_name) for t in templates]).to(device)
        embeddings = model.encode_text(texts)              # (num_templates, D)
        embeddings = F.normalize(embeddings, dim=-1)
        mean_embedding = embeddings.mean(dim=0)
        mean_embedding = F.normalize(mean_embedding, dim=-1)  # re-normalise after mean
        classifier_weights.append(mean_embedding)
    return torch.stack(classifier_weights, dim=0)          # (1000, D)

# class_names ordered by timm index (0..999)
class_names_ordered = [idx_to_name[i] for i in range(1000)]

zeroshot_weights = build_zero_shot_classifier(
    newmodel, tokenizer, class_names_ordered, IMAGENET_TEMPLATES, DEVICE
)  # (1000, D)

# ── 3. Dataset — same ImageFolder + label remapping as before ─────────────────
IMAGENET_VAL_DIR = "/home/nfm/data_prisma/imagenet_val/kaggle/input/imagenet-object-localization-challenge/ILSVRC/Data/CLS-LOC/val"
BATCH_SIZE = 256
NUM_WORKERS = 8

dataset = ImageFolder(IMAGENET_VAL_DIR, transform=preprocess)

folder_idx_to_timm_idx = {
    folder_idx: wnid_to_idx[wnid]
    for wnid, folder_idx in dataset.class_to_idx.items()
}

loader = DataLoader(dataset, batch_size=BATCH_SIZE, shuffle=False,
                    num_workers=NUM_WORKERS, pin_memory=True)

# ── 4. Zero-shot evaluation loop ──────────────────────────────────────────────
model.eval()
top1_correct = top5_correct = total = 0

with torch.no_grad():
    for images, folder_labels in tqdm(loader, desc="Zero-shot eval"):
        images = images.to(DEVICE)

        timm_labels = torch.tensor(
            [folder_idx_to_timm_idx[l.item()] for l in folder_labels],
            device=DEVICE,
        )

        # image embeddings → cosine similarity against all class text embeddings
        # image_features = model.encode_image(images)
        image_features = model(images)
        image_features = F.normalize(image_features, dim=-1)   # (B, D)

        # logits = scaled cosine sim: (B, 1000)
        logits = (image_features @ zeroshot_weights.T) * 1.0

        top1_correct += (logits.argmax(dim=1) == timm_labels).sum().item()
        top5_correct += (logits.topk(5, dim=1).indices == timm_labels.unsqueeze(1)).any(dim=1).sum().item()
        total += images.size(0)

print(f"Images   : {total:,}")
print(f"Top-1    : {100 * top1_correct / total:.2f}%")
print(f"Top-5    : {100 * top5_correct / total:.2f}%")

Zero-shot eval: 100%|██████████| 196/196 [03:43<00:00,  1.14s/it]

Images   : 50,000
Top-1    : 45.75%
Top-5    : 63.49%


org clip
Images   : 50,000
Top-1    : 65.60%
Top-5    : 87.72%

timmfinetuned
Images   : 50,000
Top-1    : 45.75%
Top-5    : 63.49%

myfinetuned
Images   : 50,000
Top-1    : 51.94%
Top-5    : 79.88%

In [ ]:
from urllib.request import urlopen
from PIL import Image
import timm
import torch

img = Image.open(urlopen(
    'https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/beignets-task-guide.png'
))

# model = timm.create_model('vit_base_patch32_clip_224.laion2b_ft_in1k', pretrained=True)
# model.to(DEVICE)
# model = model.eval()


# from vit_prisma.models.model_loader import load_hooked_model
# model_name = "vit_base_patch32_clip_224.laion2b_ft_in1k"
# model2 = load_hooked_model(
#     model_name,
#     fold_ln=True,
#     device=DEVICE
# )

# model2.eval()
# model2.to(DEVICE)

# get model specific transforms (normalization, resize)
data_config = timm.data.resolve_model_data_config(model2)
transforms = timm.data.create_transform(**data_config, is_training=False)

output = model(transforms(img).unsqueeze(0).to(DEVICE))  # unsqueeze single image into batch of 1

# top5_probabilities, top5_class_indices = torch.topk(output.softmax(dim=1) * 100, k=5)
# # print("Top 5 Predictions:")
# for prob, idx in zip(top5_probabilities[0], top5_class_indices[0]):
#     class_name = idx_to_name[idx.item()]
#     print(f"{class_name}: {prob.item():.2f}%")

2026-04-13 17:21:08 DEBUG:PIL.PngImagePlugin: STREAM b'IHDR' 16 13
2026-04-13 17:21:08 DEBUG:PIL.PngImagePlugin: STREAM b'eXIf' 41 764
2026-04-13 17:21:08 DEBUG:PIL.PngImagePlugin: STREAM b'IDAT' 817 65536


In [12]:
# 3. Compare a random attention weight matrix
# (Make sure both models are on the same device)
diff = torch.max(torch.abs(model.blocks[0].attn.W_Q - base_model.blocks[0].attn.W_Q))

print(f"Max difference in Block 0 Attention Q weights: {diff.item()}")
if diff.item() < 1e-6:
    print("Success! Your base weights were perfectly frozen during the linear probe.")
else:
    print("Warning: The base weights changed during training.")

Max difference in Block 0 Attention Q weights: 0.006670156493782997


In [4]:
base_model_name = "open-clip:laion/CLIP-ViT-B-16-laion2B-s34B-b88K"
base_model = load_hooked_model(
    base_model_name, 
    fold_ln=True, 
    device=DEVICE
)

base_model.eval()
base_model.to(DEVICE)

2026-04-13 14:00:31 INFO:root: Model 'open-clip:laion/CLIP-ViT-B-16-laion2B-s34B-b88K' is supported and passes tests.
2026-04-13 14:00:31 INFO:root: model_id download_pretrained_from_hf: laion/CLIP-ViT-B-16-laion2B-s34B-b88K
2026-04-13 14:00:32 DEBUG:urllib3.connectionpool: https://huggingface.co:443 "HEAD /laion/CLIP-ViT-B-16-laion2B-s34B-b88K/resolve/main/open_clip_config.json HTTP/1.1" 307 0
2026-04-13 14:00:32 DEBUG:urllib3.connectionpool: https://huggingface.co:443 "HEAD /api/resolve-cache/models/laion/CLIP-ViT-B-16-laion2B-s34B-b88K/7288da5a0d6f0b51c4a2b27c624837a9236d0112/open_clip_config.json HTTP/1.1" 200 0


*****Loading model 'open-clip:laion/CLIP-ViT-B-16-laion2B-s34B-b88K' of type 'VISION'


2026-04-13 14:00:32 INFO:root: model_id download_pretrained_from_hf: laion/CLIP-ViT-B-16-laion2B-s34B-b88K
2026-04-13 14:00:32 DEBUG:urllib3.connectionpool: https://huggingface.co:443 "HEAD /laion/CLIP-ViT-B-16-laion2B-s34B-b88K/resolve/main/open_clip_pytorch_model.bin HTTP/1.1" 302 0
2026-04-13 14:00:33 INFO:root: visual projection shape: torch.Size([768, 512])


LayerNorm folded.


2026-04-13 14:00:34 INFO:root: Loaded pretrained model open-clip:laion/CLIP-ViT-B-16-laion2B-s34B-b88K into HookedTransformer


HookedViT(
  (embed): PatchEmbedding(
    (proj): Conv2d(3, 768, kernel_size=(16, 16), stride=(16, 16))
  )
  (hook_embed): HookPoint()
  (pos_embed): PosEmbedding()
  (hook_pos_embed): HookPoint()
  (hook_full_embed): HookPoint()
  (ln_pre): LayerNorm(
    (hook_scale): HookPoint()
    (hook_normalized): HookPoint()
  )
  (hook_ln_pre): HookPoint()
  (blocks): ModuleList(
    (0-11): 12 x TransformerBlock(
      (ln1): LayerNorm(
        (hook_scale): HookPoint()
        (hook_normalized): HookPoint()
      )
      (ln2): LayerNorm(
        (hook_scale): HookPoint()
        (hook_normalized): HookPoint()
      )
      (attn): Attention(
        (hook_k): HookPoint()
        (hook_q): HookPoint()
        (hook_v): HookPoint()
        (hook_z): HookPoint()
        (hook_attn_scores): HookPoint()
        (hook_pattern): HookPoint()
        (hook_result): HookPoint()
      )
      (mlp): MLP(
        (hook_pre): HookPoint()
        (hook_post): HookPoint()
      )
      (hook_attn_in): Ho

## Push to HF

In [3]:
import timm
from timm.models.hub import push_to_hf_hub
from huggingface_hub import login

# 0. Log in to Hugging Face using your token
# Replace "hf_YOUR_TOKEN_HERE" with your actual Hugging Face Write token
login(token="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")

# 1. Re-create the model structure (same as training)
model = timm.create_model(
    'vit_base_patch16_clip_224.laion2b', 
    num_classes=1000, 
    pretrained=False
)

# 2. Load your trained weights from train.py's output
timm.models.load_checkpoint(
    model, 
    '/home/nfm/pytorch-image-models/output/train/20260418-015740-vit_base_patch16_clip_224_laion2b-224/model_best.pth.tar'
)

# 3. Push it directly to your Hugging Face account
push_to_hf_hub(
    model, 
    repo_id='natihash/vit_base_patch16_clip_224.laion2b_fullft',
    commit_message="Adding full fine-tuned weights for ImageNet-1k"
)

/home/nfm/prisma/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
/home/nfm/prisma/lib/python3.10/site-packages/timm/models/hub.py:4: FutureWarning: Importing from timm.models.hub is deprecated, please import via timm.models
  warnings.warn(f"Importing from {__name__} is deprecated, please import via timm.models", FutureWarning)
Processing Files (2 / 2): 100%|██████████|  693MB /  693MB, 69.3MB/s  
New Data Upload: 100%|██████████|  445MB /  445MB, 44.5MB/s  


CommitInfo(commit_url='https://huggingface.co/natihash/vit_base_patch16_clip_224.laion2b_fullft/commit/8aa4e8f66396a6f32c6f777edc8ce9d92047b36d', commit_message='Adding full fine-tuned weights for ImageNet-1k', commit_description='', oid='8aa4e8f66396a6f32c6f777edc8ce9d92047b36d', pr_url=None, repo_url=RepoUrl('https://huggingface.co/natihash/vit_base_patch16_clip_224.laion2b_fullft', endpoint='https://huggingface.co', repo_type='model', repo_id='natihash/vit_base_patch16_clip_224.laion2b_fullft'), pr_revision=None, pr_num=None)